internal/index/manager/manager.go

```diff
@@ -784,41 +785,66 @@ func (mgr *Manager) UpdateTag(name string, operation UpdateTagOperation) error {
 				if maxUsedStreamID >= mgr.nextStreamID {
 					return fmt.Errorf("unknown stream id %d", maxUsedStreamID)
 				}
-				newTag := *t
-				newTag.Matches = t.Matches.Copy()
-				for _, s := range info.markTagAddStreams {
-					newTag.Matches.Set(uint(s))
-					newTag.Uncertain.Set(uint(s))
+				// update mark streamid tag matches without parsing the definition again
+				// this is a bit hacky but it is much faster than parsing the definition of long mark tags again
+				b := strings.Builder{}
+				b.WriteString("id:")
+				for i, s := range info.markTagAddStreams {
+					if t.Matches.IsSet(uint(s)) {
+						continue
+					}
+					t.Matches.Set(uint(s))
+					t.Uncertain.Set(uint(s))
+					if i != 0 {
+						b.WriteByte(',')
+					}
+					b.WriteString(fmt.Sprintf("%d", s))
 				}
-				for _, s := range info.markTagDelStreams {
-					newTag.Matches.Unset(uint(s))
-					newTag.Uncertain.Set(uint(s))
+				if b.Len() > 3 {
+					markQuery := b.String()
+					if q, err := query.Parse(markQuery); err == nil {
+						t.Conditions = t.Conditions.Or(q.Conditions)
+					}
+					if t.definition == "id:-1" {
+						t.definition = markQuery
+					} else {
+						t.definition = fmt.Sprintf("%s,%s", t.definition, markQuery[3:])
+					}
 				}
 
-				b := strings.Builder{}
-				if newTag.Matches.IsZero() {
-					b.WriteString("id:-1")
-				} else {
-					b.WriteString("id:")
-					last := uint(0)
-					for {
-						zeros := newTag.Matches.TrailingZerosFrom(last)
-						if zeros < 0 {
-							break
-						}
-						if last != 0 {
-							b.WriteByte(',')
-						}
-						last += uint(zeros)
-						b.WriteString(fmt.Sprintf("%d", last))
-						last++
+				b.Reset()
+				b.WriteString("!id:")
+				for i, s := range info.markTagDelStreams {
+					if !t.Matches.IsSet(uint(s)) {
+						continue
 					}
+					t.Matches.Unset(uint(s))
+					t.Uncertain.Set(uint(s))
+
+					if i != 0 {
+						b.WriteByte(',')
+					}
+
+					b.WriteString(fmt.Sprintf("%d", s))
+					t.definition = t.definition[:3] + regexp.MustCompile(fmt.Sprintf("(,|^)%d(,|$)", s)).ReplaceAllString(t.definition[3:], ",")
```

Comment:
let's simplify this stuff a bit:
create a new string before the loop:
newDefinition := fmt.Sprintf(",%s,", t.definition[3:])

then, in the loop, replace the single occurence of ,N, with a ,:
newDefinition = strings.Replace(newDefinition, fmt.Sprintf(",%d,", s), ",", 1)

then, in the if b.Len() > 4:
if newDefinition == "," {
  t.definition = "id:-1"
} else {
  t.definition = fmt.Sprintf("id:%s", newDefinition[1:len(newDefinition) - 1])
}